online/src/main/scala/ai/chronon/online/Fetcher.scala

```diff
@@ -183,30 +193,46 @@ class Fetcher(val kvStore: KVStore,
             val derivationStartTs = System.currentTimeMillis()
             val joinName = internalResponse.request.name
             val ctx = Metrics.Context(Environment.JoinFetching, join = joinName)
-            val joinCodec = getJoinCodecs(internalResponse.request.name).get
-            ctx.distribution("derivation_codec.latency.millis", System.currentTimeMillis() - derivationStartTs)
-            val requestTs = internalResponse.request.atMillis.getOrElse(System.currentTimeMillis())
-            val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
-            val baseMap = internalMap ++ externalMap
-            // used for derivation based on ts/ds
-            val tsDsMap: Map[String, AnyRef] =
-              Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
-            val derivedMap: Map[String, AnyRef] = Try(
-              joinCodec
-                .deriveFunc(internalResponse.request.keys, baseMap ++ tsDsMap)
-                .mapValues(_.asInstanceOf[AnyRef])
-                .toMap) match {
-              case Success(derivedMap) => derivedMap
-              case Failure(exception) => {
+            val joinCodecTry = getJoinCodecs(internalResponse.request.name)
+            joinCodecTry match {
+              case Success(joinCodec) =>
+                ctx.distribution("derivation_codec.latency.millis", System.currentTimeMillis() - derivationStartTs)
+                val requestTs = internalResponse.request.atMillis.getOrElse(System.currentTimeMillis())
+                val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
+                val baseMap = internalMap ++ externalMap
+                // used for derivation based on ts/ds
+                val tsDsMap: Map[String, AnyRef] =
+                  Map(Constants.TimeColumn -> requestTs, "ds" -> requestDs)
+                    .mapValues(_.asInstanceOf[AnyRef])
+                    .toMap
+                val derivedMapTry: Try[Map[String, AnyRef]] = Try {
+                  joinCodec
+                    .deriveFunc(internalResponse.request.keys, baseMap ++ tsDsMap)
+                    .mapValues(_.asInstanceOf[AnyRef])
+                    .toMap
+                }
+
+                val derivedMap = derivedMapTry match {
+                  case Success(derivedMap) =>
+                    val derivedCleanedMap = derivedMap -- tsDsMap.keys
+                    derivedCleanedMap
+                  case Failure(exception) =>
+                    ctx.incrementException(exception)
+                    val derivedExceptionMap =
+                      Map("derivation_fetch_exception" -> exception.traceString.asInstanceOf[AnyRef])
+                    derivedExceptionMap
```

Comment:
@better365  Question, should we include the baseMap here? My understanding is this derivedMap defined in ResponseWithContext is the complete result after derivation, it includes both the original value(except the columns which are renamed) and the derived new columns. So even if the deriveFunc failed we still want to keep the baseMap here along with the exception message right? Pls let me know if my understanding is incorrect here.

Reply:
After discussion with the team, we will apply rename only derivations at this step. If client has a fall back logic, it will make no difference. Otherwise, it will be good to have partial responses.